File: Data Structures & Algorithms/maximum-product-subarray/submission-6.java
class Solution {
    public int maxProduct(int[] nums) {
        int ans = Integer.MIN_VALUE; // 注意答案可能是负数
        int fMax = 1;
        int fMin = 1;

        // f[i] = Math.max(f[i-1] * now, now)
        for (int now : nums) {
            int prevMax = fMax;
            fMax = Math.max(Math.max(fMax * now, fMin * now), now);
            // 因為fMax已經更新過了，所以要用之前的fMax
            fMin = Math.min(Math.min(prevMax * now, fMin * now), now);
            ans = Math.max(ans, fMax);
        }
        return ans;
    }
}

File: Data Structures & Algorithms/house-robber-ii/submission-6.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        return max(nums[0] + self.rob1(nums[2:-1]), self.rob1(nums[1::]))

    def rob1(self, nums: List[int]) -> int:
        n = len(nums)
        f0 = 0
        f1 = 0
        for i in range(n):
            tmp = max(f1, f0 + nums[i])
            f0 = f1
            f1 = tmp
        return f1            